ospd_openvas/preferencehandler.py

```diff
@@ -0,0 +1,610 @@
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+
+# pylint: disable=too-many-lines
+
+""" Prepare the preferences to be used by OpenVAS """
+
+import logging
+import uuid
+import binascii
+
+from enum import IntEnum
+from typing import Optional, Dict, List, Tuple
+from base64 import b64decode
+
+from ospd.scan import ScanCollection
+from ospd_openvas.openvas import Openvas
+from ospd_openvas.db import KbDB
+
+logger = logging.getLogger(__name__)
+
+
+OID_SSH_AUTH = "1.3.6.1.4.1.25623.1.0.103591"
+OID_SMB_AUTH = "1.3.6.1.4.1.25623.1.0.90023"
+OID_ESXI_AUTH = "1.3.6.1.4.1.25623.1.0.105058"
+OID_SNMP_AUTH = "1.3.6.1.4.1.25623.1.0.105076"
+OID_PING_HOST = "1.3.6.1.4.1.25623.1.0.100315"
+
+
+class AliveTest(IntEnum):
+    """ Alive Tests. """
+
+    ALIVE_TEST_TCP_ACK_SERVICE = 1
+    ALIVE_TEST_ICMP = 2
+    ALIVE_TEST_ARP = 4
+    ALIVE_TEST_CONSIDER_ALIVE = 8
+    ALIVE_TEST_TCP_SYN_SERVICE = 16
+
+
+def _from_bool_to_str(value: int) -> str:
+    """ The OpenVAS scanner use yes and no as boolean values, whereas ospd
+    uses 1 and 0."""
+    return 'yes' if value == 1 else 'no'
+
+
+class PreferenceHandler:
+    def __init__(
+        self, scan_id: str, kbdb: KbDB, scan_collection: ScanCollection
+    ):
+        self.scan_id = scan_id
+        self.kbdb = kbdb
+        self.scan_collection = scan_collection
+
+        self._openvas_scan_id = None
+
+        self._target_options = None
+
+    @property
+    def openvas_scan_id(self) -> str:
+        """ Create the openvas scan id and store it in the redis kb.
+        Return the openvas scan_id.
+        """
+        if self._openvas_scan_id is not None:
+            return self._openvas_scan_id
+
+        self._openvas_scan_id = str(uuid.uuid4())
+        self.kbdb.add_scan_id(self.scan_id, self._openvas_scan_id)
+
+        return self._openvas_scan_id
+
+    @property
+    def target_options(self) -> Dict:
+        """ Return target options from Scan collection """
+        if self._target_options is not None:
+            return self._target_options
+
+        self._target_options = self.scan_collection.get_target_options(
+            self.scan_id
+        )
+        return self._target_options
+
+    def get_vts_in_groups(
+        self, filters: List[str], vts_cache: Dict[str, Dict],
```

Comment:
Personally i would move the vts_cache argument to the front for all get_vt(s)_ methods but it's just a mater of taste.